src/Storages/ExternalStream/Kafka/WriteBufferFromKafka.cpp

```diff
@@ -0,0 +1,80 @@
+#include "WriteBufferFromKafka.h"
+#include <cmath>
+#include <utility>
+#include "Common/logger_useful.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int CANNOT_WRITE_TO_KAFKA;
+}
+
+WriteBufferFromKafka::WriteBufferFromKafka(const Poco::Logger * logger, RdKafkaTopicPtr topic_, size_t buffer_size)
+    : BufferWithOwnMemory<WriteBuffer>(buffer_size), log(logger), topic(std::move(topic_))
+{
+}
+
+void WriteBufferFromKafka::nextImpl()
+{
+    if (!offset())
+        return;
+
+    auto err = rd_kafka_produce(
```

Comment:
Do we need a plain produce here ? Anything we can leverage in KafkaLog or since we only have sink here a plain Kafka produce is more lightweight / straight forward ?

Reply:
Correct me if I got this wrong, based on reading the code in `KafkaLog`, the existing implementation can't not be used in this sink directly, I will need to implement something like a `KafkaWALProducer` (just like the `KafkaWALConsumer`).

Secondly, IMO, all the `KafkaSink` and `WriteBufferForKafka` classes are already a high level concept, it's totally fine to use the librdkafka library directly in these classes (these classes are already wrappers). Adding another layer looks will unneccessarily complex the code (as you can see, the logic in `KafkaSink` and `WriteBufferForKafka` are very simple and straightfoward).

So, I don't really feel adding another layer helps in this case.

However, I do think that `KafkaSource` and `KafkaSink` should be more consistent. Probably it makes sense to do some refactor with followup PRs.

I hope this makes sense.

Reply:
https://github.com/timeplus-io/proton/blob/develop/src/Storages/Streaming/StorageStream.cpp#L1740-L1792

Reply:
The current implementation doesn't retry what has been failed. We can do this in next PR if necessary. The StorageStream has some example like how to track outstanding blocks via assigning each outstanding block an ID, and if one timeout / failed, retry.

Reply:
It does retry. However, in theory, it's still possible it fails to send messages. In such cases, it will fail the checkpointing thus cause possible data duplication. Another consideration is to use transaction. But even with current implementation, based on my tests, failures like broker failures, network issues, etc. won't be an issue for us.